exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,619 @@
+package orderbook
+
+import (
+	"errors"
+	"fmt"
+	"time"
+)
+
+var errIDCannotBeMatched = errors.New("cannot match ID on linked list")
+var errCollisionDetected = errors.New("cannot insert update collision detected")
+var errAmountCannotBeZero = errors.New("amount cannot be zero")
+
+// linkedList defines a linked list for a depth level, reutilisation of nodes
+// to and from a stack.
+// TODO: Test cross link between bid and ask head nodes **node ref to head for
+// future strategy asymmetric traversal between two books
+type linkedList struct {
+	length int
+	head   *node
+}
+
+// Load iterates across new items and refreshes linked list. It creates a linked
+// list exactly the same as the item slice that is supplied, if items is of nil
+// value it will flush entire list.
+func (ll *linkedList) load(items Items, stack *stack) {
+	// Tip sets up a pointer to a struct field variable pointer. This is used
+	// so when a node is popped from the stack we can reference that current
+	// nodes' struct 'next' field and set on next iteration without utilising
+	// assignment for example `prev.next = *node`.
+	var tip = &ll.head
+	// Prev denotes a place holder to node and all of its next references need
+	// to be pushed back onto stack.
+	var prev *node
+	for i := range items {
+		if *tip == nil {
+			// Extend node chain
+			*tip = stack.Pop()
+			// Set current node prev to last node
+			(*tip).prev = prev
+			ll.length++
+		}
+		// Set item value
+		(*tip).value = items[i]
+		// Set previous to current node
+		prev = *tip
+		// Set tip to next node
+		tip = &(*tip).next
+	}
+
+	// Push has references to dangling nodes that need to be removed and pushed
+	// back onto stack for re-use
+	var push *node
+	// Cleave unused reference chain from main chain
+	if prev == nil {
+		// The entire chain will need to be pushed back on to stack
+		push = *tip
+		ll.head = nil
```

Comment:
This scenario looks like it can only exist if `items` is nil/len(0) because prev is set to tip after ensuring tip is not nil. So in this scenario, wouldn't that also make `tip` = nil too? If this then makes `push` = nil, then should this kind of item length check be at the start and return earlier?

Reply:
No, if there is already data on the linked list and we want to purge the reference and still push the entire chain on the stack *tip will still be populated. I guess I can still add a *tip == nil check and return on ln: 55